### PR TITLE
Add bit string literals for ease of encoding BIT STRINGs

### PR DESCRIPTION
**Pull request by mcy**
BIT STRING is distinct from OCTET STRING in that it has a leading byte
specifying the number of "unimportant" bits in the last octet. This adds
syntax (b`...`) similar to hex literals, which generates this prefix in
a convenient fashion. See language.txt for details.

Includes changes to ascii2der, as well as a changing the existing
der2ascii heuristic for well-formed BIT STRINGs to use the new syntax
instead.

### CODE REVIEW

File: cmd/ascii2der/scanner.go [hunk 1]
```diff
@@ -284,6 +284,51 @@ again:
...
+					if bitCount%8 == 0 {
+						bits = append(bits, 0)
+					}
+					if r == '1' {
+						bits[len(bits)-1] |= 1 << uint(bitCount%8)
+					}
+					bitCount++
+				case '|':
+					if sawPipe {
+						return token{}, &parseError{s.pos, errors.New("duplicate '")}
```
Comment (davidben) [inline cmd/ascii2der/scanner.go:312]: This and the below message reference `'` instead of `|`.

File: cmd/ascii2der/scanner.go [hunk 2]
```diff
@@ -284,6 +284,51 @@ again:
...
+					}
+
+					// Ensure that there aren't actually more than |padding| bits past the |.
```
Comment (davidben) [inline cmd/ascii2der/scanner.go:315]: Nit: Looking through the rest of the project, I don't appear to have actually done the `|` thing in comments here.

File: cmd/ascii2der/scanner.go [hunk 3]
```diff
@@ -284,6 +284,51 @@ again:
...
+					padding := (len(bits)-1)*8 - bitCount
+					if len(bitStr)-i > padding+1 {
+						return token{}, &parseError{s.pos, errors.New("more than 7 bits after '")}
```
Comment (davidben) [inline cmd/ascii2der/scanner.go:318]: The comment says `padding`, the code says `padding+1`, while the error says 7. The error string seems wrong. `padding+1` I guess is because this is really checking `len(bitStr)-i-1` to account for the pipe? That might be a bit easier to read.

File: cmd/ascii2der/scanner.go [hunk 4]
```diff
@@ -284,6 +284,51 @@ again:
...
+	case 'b':
+		if s.pos.Offset+1 < len(s.text) && s.text[s.pos.Offset+1] == '`' {
+			s.advance() // Skip the b.
+			s.advance() // Skip the `.
+			bitStr, ok := s.consumeUpTo('`')
+			if !ok {
+				return token{}, &parseError{s.pos, errors.New("unmatched `")}
+			}
+
+			// Leading byte is the number of "extra" bits at the end.
```
Comment (davidben) [inline cmd/ascii2der/scanner.go:296]: Nit: s/Leading/The leading/ for a complete sentence. I guess since the logic below describes it as padding bits, we can probably s/"extra" bits/padding bits/.

File: cmd/ascii2der/scanner_test.go [hunk 5]
```diff
@@ -224,6 +226,53 @@ indefinite long-form:2`,
...
+	// We can stick a | in the middle of a BIT STRING to add "explicit" padding.
+	{
+		"b`101010100|1010101`",
+		[]token{
+			{Kind: tokenBytes, Value: []byte{0x07, 0x55, 0xaa}},
+			{Kind: tokenEOF},
+		},
+		true,
+	},
+	// "Implicit" padding is zeroed in spite of explicit padding.
```
Comment (davidben) [inline cmd/ascii2der/scanner_test.go:263]: It took me a bit to figure out what this was saying. Perhaps:

  // If explicit padding does not end at a byte boundary, the remaining padding bits are zero.

File: cmd/ascii2der/scanner_test.go [hunk 6]
```diff
@@ -224,6 +226,53 @@ indefinite long-form:2`,
 		},
 		true,
 	},
+	// Byte-string-like BIT STRING literals are parsed correctly.
```
Comment (davidben) [inline cmd/ascii2der/scanner_test.go:229]: Maybe just "BIT STRING literals". We don't have any other kind.

File: language.txt [hunk 7]
```diff
@@ -76,6 +76,31 @@ Unicode code point."
...
+
+# This encodes as `0055`.
+b`10101010`
+
+# This encodes as `0450`.
+b`1010`
+
+# Bit string literals may also include a single | character, which marks the
+# beginning of explicit padding bytes. BER permits any bit sequence for the
+# padding bytes. However, it is an error for explicit padding to cross the octet
```
Comment (davidben) [inline language.txt:93]: s/the/an/ since there are many octet boundaries. Also looks like I used "byte" in the most of the rest of the document, so feel free to use the less silly term. :-)

File: cmd/der2ascii/writer.go [hunk 8]
```diff
@@ -355,6 +355,40 @@ func derToASCIIImpl(out *bytes.Buffer, in []byte, indent int, stopAtEOC bool) []
...
+					for i, octet := range elem.body[1:] {
+						// Last octet gets some special handling.
+						isLast := i == len(elem.body) - 2
+						for j := 0; j < 8; j++ {
+							if isLast {
+								if octet == 0 && significant <= 0 {
+									break
+								}
+								if significant == 0 {
+									bits.WriteRune('|')
```
Comment (davidben) [inline cmd/der2ascii/writer.go:377]: If the BIT STRING has correct padding, let's omit the padding bits

File: language.txt [hunk 9]
```diff
@@ -76,6 +76,31 @@ Unicode code point."
 `AbCdEf`
 
 
+# Bit string literals.
+
+# A bit string literal is a hex literal beginning with 'b'. Only 0 and 1 may
+# appear inside. Bit string literals are encoded like a BIT STRING. (Big-endian
```
Comment (davidben) [inline language.txt:82]: | can also appear inside, at which point it's not really a hex literal anymore. I also went a bit wordy in the others (probably could have picked a tighter phrasing...) because "encoded like a BIT STRING" could mean with or without the tag. Perhaps:

A backtick string beginning with 'b' is a bit string literal. 0 or 1 characters denote bits in a bit string. | characters are interpreted as below. No other characters may appear. Bit string literals emit the contents of the bit string's DER encoding as a BIT STRING. (Big-endian bit order, zero-padded, and prefixed with the number of padding bits.)

File: language.txt [hunk 10]
```diff
@@ -76,6 +76,31 @@ Unicode code point."
...
+# bit order, prefixed with a "number of padding bits" octet.)
+
+# This encodes as `0055`.
+b`10101010`
+
+# This encodes as `0450`.
+b`1010`
+
+# Bit string literals may also include a single | character, which marks the
+# beginning of explicit padding bytes. BER permits any bit sequence for the
```
Comment (davidben) [inline language.txt:92]: s/bytes/bits/

File: cmd/der2ascii/writer.go [hunk 11]
```diff
@@ -355,6 +355,40 @@ func derToASCIIImpl(out *bytes.Buffer, in []byte, indent int, stopAtEOC bool) []
 					// Emit the remaining as a DER element.
 					derToASCIIImpl(out, elem.body[1:], indent+1, false) // Adds a trailing newline.
 					addLine(out, indent, "}")
+				} else if len(elem.body) >= 1 && len(elem.body) <= 5 && elem.body[0] < 8 {
```
Comment (davidben) [inline cmd/der2ascii/writer.go:358]: I believe this will incorrectly encode {0x07} as b\`\` and break round-tripping.

File: cmd/der2ascii/writer.go [hunk 12]
```diff
@@ -355,6 +355,40 @@ func derToASCIIImpl(out *bytes.Buffer, in []byte, indent int, stopAtEOC bool) []
...
+					// Convert to a b`` literal when the leading bit is valid and the
+					// number of data octets is at most 4.
+					// 
+					// The 4 octet limit is as above: X.509 incorrectly encodes things
+					// like signatures and RSA moduli as BIT STRING.
```
Comment (davidben) [inline cmd/der2ascii/writer.go:363]: Well, it's also because overly long bit strings aren't particularly helpful on the legibility wrong.

File: cmd/ascii2der/scanner.go [hunk 13]
```diff
@@ -284,6 +284,51 @@ again:
...
+			var sawPipe bool
+			bits := []byte{0}
+			for i, r := range bitStr {
+				switch r {
+				case '0', '1':
+					if bitCount%8 == 0 {
+						bits = append(bits, 0)
+					}
+					if r == '1' {
+						bits[len(bits)-1] |= 1 << uint(bitCount%8)
```
Comment (davidben) [inline cmd/ascii2der/scanner.go:307]: I think you've got the bits backwards. Ditto with writer.go. The bits are ordered from 8 to 1 (X.690-202102, 8.6.2.1) or MSB to LSB. This is doing the LSB first.

You can see this in the `101010100101` test below. The test says {0x04, 0x??, 0x0?}, but a DER BIT STRING on a nibble boundary will always look like {0x04, 0x??, 0x?0}.

**Review by davidben — CHANGES_REQUESTED** [review]

File: cmd/ascii2der/scanner.go (hunk 1 above)
Comment (mcy) [reply cmd/ascii2der/scanner.go:312]: Thought I got them all. =/

File: cmd/ascii2der/scanner.go (hunk 3 above)
Comment (mcy) [reply cmd/ascii2der/scanner.go:318]: Added some more variables to clear things up. I wasn't deeply happy with how I organized this code.

File: language.txt (hunk 9 above)
Comment (mcy) [reply language.txt:82]: Much better! I clamped down the wording on the other paragraph too.

File: cmd/der2ascii/writer.go (hunk 8 above)
Comment (mcy) [reply cmd/der2ascii/writer.go:377]: Not sure I follow. I think this code correctly elides any trailing padding bits that are zero? For example, the text case
```go
	{
		[]byte{0x03, 0x05, 0x01, 0x30, 0x80, 0x00, 0x00},
		"BIT_STRING { b`0000110000000001000000000000000` }\n",
	},
```
would emit
```
b`0000110000000001000000000000000|0`
```
if this was not the case.

(Maybe I should include a better test case. Thoughts?)

File: cmd/ascii2der/scanner.go (hunk 13 above)
Comment (mcy) [reply cmd/ascii2der/scanner.go:307]: ...yup. I made sure to try it the right way around and still got it wrong. Thanks for catching this. =/

**Comment by davidben** [comment]
Hrm, did you forget to upload the new iteration of the PR, or am I just using GitHub wrong?

**Comment by mcy** [comment]
Apparently I did? Should be fixed now.

File: cmd/ascii2der/scanner.go [hunk 14]
```diff
@@ -284,6 +284,53 @@ again:
...
+			s.advance() // Skip the `.
+			bitStr, ok := s.consumeUpTo('`')
+			if !ok {
+				return token{}, &parseError{s.pos, errors.New("unmatched `")}
+			}
+
+			// The leading byte is the number of "extra" bits at the end.
+			var bitCount int
+			var sawPipe bool
+			bits := []byte{0}
```
Comment (davidben) [inline cmd/ascii2der/scanner.go:299]: Nitpick: `bits` made me read this as a `[]bool` when it's not. :-) `bytes` is a module name, so how about `value`?

File: cmd/ascii2der/scanner.go [hunk 15]
```diff
@@ -284,6 +284,53 @@ again:
...
+			for i, r := range bitStr {
+				bitsRemaining := (len(bits)-1)*8 - bitCount
+				switch r {
+				case '0', '1':
+					if bitCount%8 == 0 {
+						bits = append(bits, 0)
+					}
+					if r == '1' {
+						bits[len(bits)-1] |= 1 << uint(bitsRemaining)
```
Comment (davidben) [inline cmd/ascii2der/scanner.go:308]: I think this still isn't right. :-) You've actually ordered the bits by: `1<<0`, `1<<7`, `1<<6`, `1<<5`, `1<<4`, `1<<3`, `1<<2`, `1<<1`. It's hard to tell because all your tests alternate bits. :-) Maybe we should add some very basic tests <code>b\`1\`</code>, <code>b\`11\`</code>, through, <code>b\`1111111\`</code>. I find that a lot easier to tell what's going on in terms of bit ordering. (Maybe something to be said for some long-ish random input too?)

Part of what's going on is that `bitsRemaining` is computed before `bitCount` and the other part is it should be `bitsRemaining - 1`. But perhaps it'd be easier just to write:
```
   if bitCount%8 == 0 {
     bits = append(bits, 0)
   }
   if r == '1' {
     // Or maybe we keep the byte index at len(bits)-1? Don't feel very strongly... pairing /8 with
     // %8 just feels tidy.
     bits[bitCount/8 + 1] |= 1 << uint(7 - bitCount%8)
   }
```

Then we can move `bitsRemaining` into the `|` case.

File: cmd/der2ascii/writer.go [hunk 16]
```diff
@@ -355,6 +355,44 @@ func derToASCIIImpl(out *bytes.Buffer, in []byte, indent int, stopAtEOC bool) []
...
+				} else if len(elem.body) == 1 && elem.body[0] == 0 {
+					addLine(out, indent, fmt.Sprintf("%s b`` }", header))
+				} else if len(elem.body) > 1 && len(elem.body) <= 5 && elem.body[0] < 8 {
+					// Convert to a b`` literal when the leading bit is valid and the
+					// number of data octets is at most 4.
+					//
+					// The 4 octet limit is as above: X.509 incorrectly encodes things
+					// like signatures and RSA moduli as BIT STRING, and because
+					// overly-long bit strings don't tend to be particularly helpful for
+					// readability.
```
Comment (davidben) [inline cmd/der2ascii/writer.go:367]: Nit: s/overly-long bit strings don't tend to be particularly helpful for readability/long bit string literals are not readable/

I suggest this because "don't tend to be particularly helpful for readability" sounds _exactly_ like something I would write. My first pass on everything is too verbose and I always take a second pass to trim it down. :-)

File: cmd/der2ascii/writer.go (hunk 8 above)
Comment (davidben) [reply cmd/der2ascii/writer.go:381]: Ah sorry, I missed the `break`.

Optional: Do you think it would be clearer to include all the padding bits if any are non-zero? That is, drop the `octet == 0 && significant <= 0` block and instead add `if octet == 0 { break }` right before `bits.WriteRune('|')`.

I could go either way on this, but my thinking is that, of the representations of `{0x07, 0x80}`, I'd probably order them in preference by:
1. <code>b\`1\`</code> — this captures the logical value
2. <code>b\`1|0000000\`</code> — this captures what's been encoded after you pad it
3. <code>b\`1|0\`</code> — this is... something

And so it seems, by extension, `{0x07, 0xc0}` would be better described by <code>b\`1|1000000\`</code> than <code>b\`1|1\`</code> because those zeros aren't any less of padding bits than the 1. It's just that the 1 is the last "wrong" padding bit.

But I could also be persuaded by "let's output the shortest description". Thoughts?

File: cmd/der2ascii/writer.go [hunk 17]
```diff
@@ -355,6 +355,44 @@ func derToASCIIImpl(out *bytes.Buffer, in []byte, indent int, stopAtEOC bool) []
...
+						isLast := i == len(elem.body)-2
+						for j := 0; j < 8; j++ {
+							if isLast {
+								if octet == 0 && significant <= 0 {
+									break
+								}
+								if significant == 0 {
+									bits.WriteRune('|')
+								}
+								significant--
```
Comment (davidben) [inline cmd/der2ascii/writer.go:383]: Nit: Rather than counting down `significant`, maybe just compare `j` against `significant`? It's a hair shorter, and it feels mildly weird to set up a counter way at the begging and only touch it at `isLast`.

File: cmd/der2ascii/writer_test.go [hunk 18]
```diff
@@ -146,7 +158,12 @@ var derToASCIITests = []convertFuncTest{
 	// Empty BIT STRINGs do not emit extra whitspace.
 	{
 		[]byte{0x03, 0x01, 0x00},
-		"BIT_STRING { `00` }\n",
+		"BIT_STRING { b`` }\n",
+	},
+	// "Empty" BIT STRINGs that assert they have padding emitted as bytes.
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:163]: Perhaps just "Invalid BIT STRINGs are encoded as bytes" and also include one that claims 8 padding bits.

File: cmd/der2ascii/writer.go [hunk 19]
```diff
@@ -355,6 +355,44 @@ func derToASCIIImpl(out *bytes.Buffer, in []byte, indent int, stopAtEOC bool) []
 					// Emit the remaining as a DER element.
 					derToASCIIImpl(out, elem.body[1:], indent+1, false) // Adds a trailing newline.
 					addLine(out, indent, "}")
+				} else if len(elem.body) == 1 && elem.body[0] == 0 {
+					addLine(out, indent, fmt.Sprintf("%s b`` }", header))
+				} else if len(elem.body) > 1 && len(elem.body) <= 5 && elem.body[0] < 8 {
+					// Convert to a b`` literal when the leading bit is valid and the
+					// number of data octets is at most 4.
+					//
+					// The 4 octet limit is as above: X.509 incorrectly encodes things
```
Comment (davidben) [inline cmd/der2ascii/writer.go:364]: Oh huh. Between the comments, the code, the test, and the spec, we seem a bit undecided about the ordering here. Suppose some input is both made of elements _and_ is a valid, short bit string.

* The spec says bit string literal wins.
* The code says elements wins.
* This comment is written as if bit string literal wins. (If elements wins, the X.509 thing doesn't figure into the length limit.)
* The test is written as if elements win, both in expectations and because the various "BIT STRINGs are encoded normally" tests only make sense if elements win. (If bit strings win, I think we'd want them to use long inputs, to test that the elements check is sensitive to non-zero padding, etc.)

I don't feel strongly about which one. I think both are plausible. (You could argue that a short bit string is probably a bitmask because it's short and anything X.509 sticks in BIT STRING is long. You could also argue that the `isMadeOfElements` heuristic seems to be reasonably good in practice and so we may as well let it win.)

File: cmd/der2ascii/writer_test.go [hunk 20]
```diff
@@ -131,12 +131,24 @@ var derToASCIITests = []convertFuncTest{
 	// BIT STRINGs are encoded normally if the contents are not an element.
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:131]: Probably worth revising these comments because "normally" could mean one of two things now. (Though see comment about elements vs. short bit string ordering. We should resolve that first.)

File: cmd/der2ascii/writer_test.go [hunk 21]
```diff
@@ -131,12 +131,24 @@ var derToASCIITests = []convertFuncTest{
...
 	{
 		[]byte{0x03, 0x03, 0x00, 0x00, 0x00},
-		"BIT_STRING { `00` `0000` }\n",
+		"BIT_STRING { b`0000000000000000` }\n",
 	},
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:135]: I believe you got the bit order correct this time, but let's add some 1, 11, etc., tests now just to be sure. :-)

**Review by davidben — COMMENTED** [review]
Sorry about the delay. Hopefully this is the last round before we get the darn bit order right!

File: cmd/ascii2der/scanner.go (hunk 14 above)
Comment (mcy) [reply cmd/ascii2der/scanner.go:299]: It's a `vector<bool>`. =P

File: cmd/ascii2der/scanner.go (hunk 15 above)
Comment (mcy) [reply cmd/ascii2der/scanner.go:308]: That was surprisingly fussy. Big endian bits is... quite unintuitive.

File: cmd/der2ascii/writer.go (hunk 19 above)
Comment (mcy) [reply cmd/der2ascii/writer.go:364]: I think the comment is just worded poorly and I didn't think about the spec too hard. Elements should always win if possible. I'll update the comment and the spec to make that clear.

File: cmd/der2ascii/writer_test.go (hunk 20 above)
Comment (mcy) [reply cmd/der2ascii/writer_test.go:131]: s/normally/bit string literals, does that catch what you meant?

File: cmd/der2ascii/writer_test.go (hunk 21 above)
Comment (mcy) [reply cmd/der2ascii/writer_test.go:135]: We did get it right, as the new tests confirm. =)

File: language.txt [hunk 22]
```diff
@@ -76,6 +76,32 @@ Unicode code point."
...
+# Bit string literals.
+
+# A backtick string beginning with 'b' is a bit string literal. 0 or 1
+# characters denote bits in a bit string. | characters are interpreted as below.
+# No other characters may appear. The emit the contents of the bit string's DER
+# encoding as a BIT STRING. (Big-endian bit order, prefixed with the number of
+# trailing padding bits)
+
+# This encodes as `0055`.
+b`10101010`
```
Comment (davidben) [inline language.txt:88]: You know what I'm going to say... :-)

(s/5/a/ in all these examples, or use 0101... instead of 1010...)

I will say, the bit order is annoying to write in code, but it's actually very convenient as a human, because we write everything big-endian. Leading byte aside, <code>b\`10101010\`</code> is just the byte `0b10101010`. Group them by fours to get hex, and `0b1010 = 0xa`.

File: language.txt [hunk 23]
```diff
@@ -288,11 +314,22 @@ SEQUENCE `aabbcc`
...
-#       remainder as a raw byte string. Otherwise, apply step g to encode the
-#       body. This is to account for X.509 incorrectly using BIT STRING
-#       instead of OCTET STRING for SubjectPublicKeyInfo and signatures.
+#    d. If the tag is BIT STRING, try the following until one succeeds:
+#       
+#       i.   If the body is exactly `00`, encode as b``.
+#
+#       ii.  If the first byte is zero, and the body happens to be valid BER,
+#            encode as `00` followed by the results of step g. This, and step
+#            iv, account for X.509 incorrectly using BIT STRING instead of
```
Comment (davidben) [inline language.txt:323]: Does step iv have anything to do with this? (Although see comment below.)

File: language.txt [hunk 24]
```diff
@@ -288,11 +314,22 @@ SEQUENCE `aabbcc`
...
+#            OCTET STRING for SubjectPublicKeyInfo and signatures.
+#
+#       iii. If the first byte is at most 7, and there are between one and four
+#            bytes after it, encode it as a bit string literal.
+#
+#       iv.  If the first byte is at most 7, encode that byte as a single-byte
+#            byte string, and then encode the rest of it as a byte string.
+#
+#       v.   Give up and encode it as a single byte string.
```
Comment (davidben) [inline language.txt:332]: I think we can fold (i) into (iii). I also wonder if it's better to talk about whether it's a valid bit string.

```
d. If the tag is a BIT STRING:

   i.   If the body is a valid bit string, contains a whole number of bytes, and
        may be parsed as a series of BER elements with no trailing data, encode
        as `00` followed by recursing into the body as in step g. This accounts
        for X.509 incorrectly using BIT STRING instead of OCTET STRING for
        SubjectPublicKeyInfo and signatures.

   ii.  If the body is a valid bit string with at most 32 bits, encode as a bit
        string literal. If any padding bits are non-zero, they are encoded
        explicitly.

   iii. If the body is a valid bit string with more than 32 bits, encode as a
        pair of hex literals, containing the initial byte and the data bytes.

   iv.  Otherwise, the body is not a valid bit string. Encode as a single
        hex literal.
```

File: cmd/der2ascii/writer_test.go [hunk 25]
```diff
@@ -128,25 +128,80 @@ var derToASCIITests = []convertFuncTest{
...
+		"BIT_STRING { b`1|1000000` }\n",
+	},
+	// BIT STRINGs are encoded as bit string literals if the leading byte is
+	// non-zero. Note that the bit string is precisely 31 bits wide.
 	{
 		[]byte{0x03, 0x05, 0x01, 0x30, 0x80, 0x00, 0x00},
-		"BIT_STRING { `01` `30800000` }\n",
+		"BIT_STRING { b`0011000010000000000000000000000` }\n",
+	},
+	// BIT STRINGs are encoded bit string literals if the leading byte is non-zero
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:176]: "encoded **as** bit string literals:", same thing about leading byte, and s/have/has/. How about just:

// The above, but with non-trivial padding.

(I think this is testing the same thing as the `1|1000000` case above.)

File: cmd/der2ascii/writer_test.go [hunk 26]
```diff
@@ -128,25 +128,80 @@ var derToASCIITests = []convertFuncTest{
...
+	// and have non-trivial padding.
+	{
+		[]byte{0x03, 0x05, 0x01, 0x30, 0x80, 0x00, 0xff},
+		"BIT_STRING { b`0011000010000000000000001111111|1` }\n",
+	},
+	// BIT STRINGs with more than four components are hex-decoded instead.
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:182]: hex-encoded

File: cmd/der2ascii/writer_test.go [hunk 27]
```diff
@@ -128,25 +128,80 @@ var derToASCIITests = []convertFuncTest{
...
+		[]byte{0x03, 0x02, 0x01, 0x100 - (1 << 1)},
+		"BIT_STRING { b`1111111` }\n",
+	},
+	// BIT STRINGs with nonzero padding always emit the entire final byte.
+	{
+		[]byte{0x03, 0x02, 0x07, 0xc0},
+		"BIT_STRING { b`1|1000000` }\n",
+	},
+	// BIT STRINGs are encoded as bit string literals if the leading byte is
+	// non-zero. Note that the bit string is precisely 31 bits wide.
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:171]: Not sure "the leading byte is non-zero" is true... maybe "if the body is valid and at most 32 bits"?

File: cmd/der2ascii/writer_test.go [hunk 28]
```diff
@@ -128,25 +128,80 @@ var derToASCIITests = []convertFuncTest{
...
+	{
+		[]byte{0x03, 0x05, 0x01, 0x30, 0x80, 0x00, 0xff},
+		"BIT_STRING { b`0011000010000000000000001111111|1` }\n",
+	},
+	// BIT STRINGs with more than four components are hex-decoded instead.
+	{
+		[]byte{0x03, 0x06, 0x01, 0x30, 0x80, 0xaa, 0x55, 0xaa},
+		"BIT_STRING { `01` `3080aa55aa` }\n",
+	},
+	// Empty BIT STRINGs do not emit extra whitespace.
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:187]: This comment was originally referring to us not emitting an extra space, in the old form, we needed to avoid an extra space when the second literal was empty. How about just fold this in with the 1, 11, 111, train to save having to write a comment. :-)

**Review by davidben — COMMENTED** [review]
Last comments. :-)

File: language.txt (hunk 23 above)
Comment (mcy) [reply language.txt:323]: Honestly, I have no idea. It took me way too many tries to write this paragraph.

File: language.txt (hunk 24 above)
Comment (mcy) [reply language.txt:332]: That is way better.

File: cmd/der2ascii/writer.go (hunk 8 above)
Comment (mcy) [reply cmd/der2ascii/writer.go]: Went with `` b`1|1000000` ``. That one requires the least thinking, I think. `` b`1` `` is still better than `` b`1|0000000` `` though.

File: cmd/der2ascii/writer.go [hunk 29]
```diff
@@ -355,6 +355,37 @@ func derToASCIIImpl(out *bytes.Buffer, in []byte, indent int, stopAtEOC bool) []
 					// Emit the remaining as a DER element.
 					derToASCIIImpl(out, elem.body[1:], indent+1, false) // Adds a trailing newline.
 					addLine(out, indent, "}")
+				} else if len(elem.body) == 1 && elem.body[0] == 0 {
+					addLine(out, indent, fmt.Sprintf("%s b`` }", header))
+				} else if len(elem.body) > 1 && len(elem.body) <= 5 && elem.body[0] < 8 {
+					// Convert to a b`` literal when the leading bit is valid and the
```
Comment (davidben) [inline cmd/der2ascii/writer.go:361]:
```suggestion
					// Convert to a b`` literal when the leading byte is valid and the
```

File: cmd/der2ascii/writer_test.go [hunk 30]
```diff
@@ -128,25 +128,87 @@ var derToASCIITests = []convertFuncTest{
...
+	},
+	// BIT STRINGs are encoded as bit string literals if the they are at most 32
+	// bits.
 	{
 		[]byte{0x03, 0x05, 0x01, 0x30, 0x80, 0x00, 0x00},
-		"BIT_STRING { `01` `30800000` }\n",
+		"BIT_STRING { b`0011000010000000000000000000000` }\n",
+	},
+	// BIT STRINGs are encoded bit string literals if the leading byte is non-zero
+	// and have non-trivial padding.
```
Comment (davidben) [inline cmd/der2ascii/writer_test.go:189]:
```suggestion
	// The above, but with non-trivial padding.
```

File: language.txt [hunk 31]
```diff
@@ -288,11 +314,24 @@ SEQUENCE `aabbcc`
...
+#
+#       ii.  If the body is a valid bit string with at most 32 bits, encode as a
+#            bit string literal. If any padding bits are non-zero, they are
+#            encoded explicitly.
+#
+#       iii. If the body is a valid bit string with more than 32 bits, encode as
+#            apair of hex literals, containing the initial byte and the data
+#            bytes.
+#
+#       v.   Otherwise, the body is not a valid bit string. Encode as a single
```
Comment (davidben) [inline language.txt:333]:
```suggestion
#       iv.  Otherwise, the body is not a valid bit string. Encode as a single
```

**Review by davidben — APPROVED** [review]
Finally figured out how suggestions in GitHub worked.
